contracts/truefi2/TimeAveragedBaseRateOracle.sol

```diff
@@ -0,0 +1,149 @@
+        // get the last timestamp written into the buffer
+        uint256 lastWritten = totalsBuffer.timestamps[totalsBuffer.currIndex];
+        return block.timestamp >= lastWritten.add(cooldownTime);
+    }
+
+    /**
+     * @dev Helper function to get contents of the totalsBuffer
+     */
+    function getTotalsBuffer()
+        public
+        view
+        returns (
+            uint256[BUFFER_SIZE] memory,
+            uint256[BUFFER_SIZE] memory,
+            uint16
+        )
+    {
+        return (totalsBuffer.runningTotals, totalsBuffer.timestamps, totalsBuffer.currIndex);
+    }
+
+    /**
+     * @dev Update the totalsBuffer:
+     * Gets current variable borrow apy from a collateralized lending protocol
+     * for chosen asset and writes down new total running value.
+     * If the buffer is filled overwrites the oldest value
+     * with a new one and updates its timestamp.
+     */
+    function update() public offCooldown {
+        uint16 _currIndex = totalsBuffer.currIndex;
+        uint16 nextIndex = (_currIndex + 1) % bufferSize();
+        uint256 apy = spotOracle.getRate(asset);
+        uint256 nextTimestamp = block.timestamp;
+        uint256 dt = nextTimestamp.sub(totalsBuffer.timestamps[_currIndex]);
+        totalsBuffer.runningTotals[nextIndex] = totalsBuffer.runningTotals[_currIndex].add(apy.mul(dt));
+        totalsBuffer.timestamps[nextIndex] = nextTimestamp;
+        totalsBuffer.currIndex = nextIndex;
+    }
+
+    /**
+     * @dev Average apy is calculated by taking
+     * the time-weighted average of the borrowing apys.
+     * Essentially formula given below is used:
+     *
+     *           sum_{i=1}^{n} v_i * (t_i - t_{i-1})
+     * avgAPY = ------------------------------------
+     *                      t_n - t_0
+     *
+     * where v_i, t_i are values of the apys and their respective timestamps.
+     * Index n corresponds to the most recent values and index 0 to the oldest ones.
+     *
+     * To avoid costly computations in a loop an optimization is used:
+     * Instead of directly storing apys we store calculated numerators from the formula above.
+     * This gives us most of the job done for every calculation.
+     *
+     * @param numberOfValues How many values of totalsBuffer should be involved in calculations.
+     * @return Average apy.
+     */
+    function calculateAverageAPY(uint16 numberOfValues) public view returns (uint256) {
+        require(numberOfValues > 0, "TimeAveragedBaseRateOracle: Number of values should be greater than 0");
+        require(numberOfValues < bufferSize(), "TimeAveragedBaseRateOracle: Number of values should be less than buffer size");
+
+        uint16 _currIndex = totalsBuffer.currIndex;
+        uint16 startIndex = (_currIndex + bufferSize() - numberOfValues) % bufferSize();
+
+        require(totalsBuffer.timestamps[startIndex] > 0, "TimeAveragedBaseRateOracle: There are fewer values stored than required");
```

Comment:
So the problem is: we can't use the oracle for a month after deployment. Maybe we should startIndex to 0 in this case? This will only be happening during the first month anyway.

Reply:
Hmm. Yeah, I'm okay with that. We would be exposed to a little more volatility in the first month, but at least the oracle would work instead of reverting.

Reply:
We can check this in and decide whether to make that change in a separate PR. Most of this PR is ready and doesn't need to wait on this decision.